File: python/tests/test_timing.py
from __future__ import annotations

import os
import time

import numpy as np
import pytest

from eclosure import closedBY, closedeBH


def _skip_unless_enabled() -> None:
  if os.getenv("ECLOSURE_RUN_TIMING_TESTS") != "1":
    pytest.skip("Set ECLOSURE_RUN_TIMING_TESTS=1 to run timing benchmarks")


def test_timing_cebh_on_1e6_evalues():
  _skip_unless_enabled()

  rng = np.random.default_rng(1)
  evalues = rng.exponential(scale=5.0, size=1_000_000)

  start = time.perf_counter()
  res = closedeBH(evalues, alpha=0.05)
  elapsed = time.perf_counter() - start

  assert isinstance(res, (int, np.integer))
  assert res >= 0
  print(f"\nclosedeBH timing (1e6 e-values): {elapsed:.3f} seconds")


def test_timing_cby_on_1e5_pvalues():
  _skip_unless_enabled()

  rng = np.random.default_rng(1)
  pvalues = rng.random(100_000) ** 4

  start = time.perf_counter()
  res = closedBY(pvalues, alpha=0.05)
  elapsed = time.perf_counter() - start

  assert isinstance(res, (int, np.integer))
  assert res >= 0
  print(f"\nclosedBY timing (1e5 p-values): {elapsed:.3f} seconds")

File: r/eclosure/tests/testthat/test_timing.R
test_that("timing benchmark: closedeBH on 1e6 e-values", {
  skip_if(
    Sys.getenv("ECLOSURE_RUN_TIMING_TESTS") != "1",
    "Set ECLOSURE_RUN_TIMING_TESTS=1 to run timing benchmarks"
  )

  set.seed(1)
  e <- rexp(1e6, rate = 1 / 5)

  timing <- system.time({
    res <- closedeBH(e)
  })

  expect_true(is.numeric(res) || is.integer(res))
  expect_length(res, 1)
  expect_gte(res, 0)
  message(sprintf(
    "closedeBH timing (1e6 e-values): %.3f seconds",
    timing[["elapsed"]]
  ))
})


test_that("timing benchmark: closedBY on 1e5 p-values", {
  skip_if(
    Sys.getenv("ECLOSURE_RUN_TIMING_TESTS") != "1",
    "Set ECLOSURE_RUN_TIMING_TESTS=1 to run timing benchmarks"
  )

  set.seed(1)
  p <- runif(1e5)^4

  timing <- system.time({
    res <- closedBY(p)
  })

  expect_true(is.numeric(res) || is.integer(res))
  expect_length(res, 1)
  expect_gte(res, 0)
  message(sprintf(
    "closedBY timing (1e5 p-values): %.3f seconds",
    timing[["elapsed"]]
  ))
})
